### PR TITLE
Fix potential huge allocations when reading TermsQueryBuilder.BinaryValues from the network

### PR DESCRIPTION
**Pull request by original-brownbear**
We should be reading a single `BytesReference` (that would be backed by a single large `byte[]`) here when we care about the individual values in the list only. Without breaking the behavior of only serializing once when sending to multiple targets this change:
* lazy serializes as needed and keeps the original terms, so we don't needlessly go through serialization in e.g. a single node situation or or requests that are handled on the coordinator directly (concurrency should be fine here, we serialize on the same thread in practice and should we ever not be on the same thread at all times this will worst case lead to serializing multiple times).
* stops allocating a potentially huge byte[] when receiving these things over the wire
* did not touch the wire format even though we could technically save 2 vint here, maybe in a follow-up but figured I'd keep this one short

follow up to #105220

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java [hunk 1]
```diff
@@ -490,64 +493,45 @@ public void clear() {
...
+            return convert ? Iterators.map(iter, AbstractQueryBuilder::maybeConvertToBytesRef) : iter;
         }
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
-            out.writeBytesReference(valueRef);
+            out.writeBytesReference(asBytes());
+        }
+
+        private BytesReference asBytes() {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:510]: this caching makes me nervous. Is there never a situation where the same query builder instance is shared across different threads? Sounds like an edge case at least, but couldn't it theoretically happen?

**Review by javanna — COMMENTED** [review]
Can you check that we have coverage in TermsQueryBuilderTests around serialization code? I know the assumption is that nothing changes, but it's easy to miss something with these changes. Do we need to introduce a new version based unit test to double check?

File: server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:510]: > Is there never a situation where the same query builder instance is shared across different threads? Sounds like an edge case at least, but couldn't it theoretically happen?

I think practically this is not something that happens right now. But even if it were to happen, we always build the bytes before we assign them. So worst case, we would serialise more often than we need to which is IMO much preferable to introducing a volatile that adds yet more hard to understand memory effects?

**Comment by original-brownbear** [comment]
> Can you check that we have coverage in TermsQueryBuilderTests around serialization code?

We do have that coverage :) Made sure of that and also debugged around it a little to understand what's going on exactly when I tracked this down.

> Do we need to introduce a new version based unit test to double check?

I don't think that would help, we don't have any versioned serialisation code here. I reread this a bunch of times though and it's rather simple logic here and I'd have to make the same mistake on both ends of the serialisation to break the format :) I think we're good here?

**Review by javanna — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Luca!

**Pull request merged**